README.md
# MMM-CTA

![Example of MMM-CTA](./screenshots/MMM-CTA.png)

This [MagicMirror Module][mm] is designed to display incoming bus and train times for the CTA (Chicago Transit Authority).  The module will display up to the minute data for bus and train arrival times.

## Get your CTA API keys!

You need to obtain API keys to access the live data to update your mirror:

Bus API:  http://www.transitchicago.com/developers/bustracker.aspx

Train API:  http://www.transitchicago.com/developers/traintrackerapply.aspx

The bus tracker API key can be obtained immediately, the train tracker key can take a few days to register.  Please reference this github since it's already approved maybe it will help speed things along for you.

## Installation

In your terminal, go to your MagicMirror's Module folder:

```bash
cd ~/MagicMirror/modules
```
Clone this repository:
```bash
git clone https://github.com/NateDee/MMM-CTA.git
```
Configure the module in your config.js file.


## Using the module

To use this module, add it to the modules array in your config.js file.

```js
modules: [
     {
	module: 'MMM-CTA',
	position: 'bottom_left',
	config: {
		updateTime: 60000, // 1 minute, the API does not update much more often so going below this is unnecessary
		ctaApiKey: 'you-bus-APIkey',
		busStopName: '',  // String value, Name your bus stop
		stopId: 561, // Bus station ID: Chicago and Milwaukee example; go to http://www.transitchicago.com/riding_cta/systemguide/default.aspx to find your stop ID
		maxResult: 4,  // The maximum number of incoming bussesy you want to display for bus stops
		ctaApiKeyTrain: 'your-train-APIkey',
		trainStopName: '',  //String value, name your train stop
		trainStationID: 41410, //Train station ID:  Chicago Blue line example; http://www.transitchicago.com/developers/ttdocs/default.aspx#_Toc296199909
		maxResultTrain: 5, // Max number of incoming trains to disply
		moduleInstance: 1, // To run multiple instances of this module
	},
     }
]
```

## Configuration Options

Option|Description
------|-----------
`updateTime`|How often the requests should update; 60 * 1000 = 1 minute.<br/>**Expected Value type:** `int`
`ctaApiKey`|Your CTA Bus API key.<br/>**Expected Value type:** `string`
`busStopName`|Name you would like to give your bus stop.<br/>**Expected Value type:** `string`
`stopId`|The number of your stopID, find a stop ID:<br/>http://www.transitchicago.com/riding_cta/systemguide/default.aspx<br/>For help finding your stopID, see the FAQ below.<br/>**Expected Value type:** `int`
`maxResult`|The maximum number of bus results your station should display.<br/>**Expected Value type:** `int`
`ctaApiKeyTrain`|Your CTA Traing API key.<br/>**Expected Value type:** `string`
`trainStopName`|Name you would like to give your train stop.<br/>**Expected Value type:** `string`
`trainStationID`|Number of L station, find a station:<br/>http://www.transitchicago.com/developers/ttdocs/default.aspx#_Toc296199909<br/>**Expected Value type:** `int`
`maxResultTrain`|The maximum number of train results to display.<br/>**Expected Value type:** `int`
`moduleInstance`|If running multiple instances, add a new instance to your config.js and increment this value by 1 for each additional instance.<br/>**Expected Value type:** `int`


## FAQ
- I can't find my bus stop ID
	
	Follow the link in "Configuration Options", use the drop down menu to find your bus route.  Under "Route Information", click the "Bus Stop List" link, which provides a table of StopID's matching the stop name and direction of travel.
	
- I can't find my train station ID

	Follow the link in "Configuration Options", scroll down to the section titled "Parent Stop ID Quick Reference" and find your stationID.  The table is labeled as StopID, but these are actually the StationID's.
	
- I want more or different config options
	
	Create an issue here please.

[mm]: https://github.com/MichMich/MagicMirror


changelog.md
Changelog for changes made and things to do

# Version 0.0.1

- To do: fix node helper nested request strips, add case for if train gets error after bus

- To do: Create conditionals for requests, based on if API keys are set for bus only, train only, or bus and train

- To do: Create full readme file that matches MMM specifications

- To do: After fixes, proof for other things need to change; if nothing else to change release to community

- To do: Add multi-instance option:  Can be done by setting an instance variable in config, and then using socketNotification with instance ID

- To do: Fix max result of bus and train

# Version 0.0.2

- Addressed all issues under version 0.0.1 and issue #1

- To do:  Add bus coloring options based on 'L' train color


MMM-CTA.js
Module.register("MMM-CTA", {
	defaults: { // Start with the information needed for a single stop
		busStopName: null,
		// stationId: null,
		stopId: null,
		maxResult: null,
		maxResultTrain: null,
		routeCode: null,
		ctaApiKey: null,
		updateTime: 60000, // 1 minute
		busUrl: 'http://www.ctabustracker.com/bustime/api/v2/getpredictions', // deleted http:
		// initialLoadDelay: 0, // This is obsolete, loaded first call by start()
		trainUrl: 'http://lapi.transitchicago.com/api/1.0/ttarrivals.aspx',
		ctaApiKeyTrain: null,
		trainStationID: null,
		trainStopName: null,
		moduleInstance: 1,
	},

	// requireVersion: 

	getStyles: function() {
		return ["font-awesome.css"];
	},

/*	loader: function() {
		var loader = document.createElement("div");
		loader.innerHTML = ("LOADING CTA TRACKER");
		loader.className = "small dimmed";
		return loader;
		this.loaded = false;
		Log.log("Loader run.")
	},
*/
	header: function() {
		var header = document.createElement("header");
		header.innerHTML = this.config.busStopName;
		return header;
	},

	start: function() {
		Log.info("Starting module: " + this.name);
		// Prevent API abuse, bus tracker limits 10,000/day
		if (this.updateTime < 60000) {
			this.updateTime = 60000 // Every 15 sec (overkill as they don't update API that often);
		};
		this.loaded = false;
		this.instanceData = "MMM-CTA-DATA" + this.config.moduleInstance;
		// this.instanceReq = "CTA-REQUEST" + this.config.moduleInstance;
		// Initial run to start;
		if (this.config.moduleInstance === 1) {
			this.apiRequest(this);
			this.scheduleUpdate();
		} else if (this.config.moduleInstance > 1) {
			setTimeout(this.apiRequest, 4000 * this.config.moduleInstance, this);
			setTimeout(this.scheduleUpdate, 4000 * this.config.moduleInstance);
		};
	
	},

	apiRequest: function(self) {
		// Variables needed for API request, see CTA devel documentation
		var request = {
			// mapid: self.config.stationId,
			stpid: self.config.stopId,
			maxRes: self.config.maxResult,
			maxResTrain: self.config.maxResultTrain,
			rt: self.config.routeCode,
			key: self.config.ctaApiKey,
			urlBus: self.config.busUrl,
			updateInterval: self.config.updateTime,
			urlTrain: self.config.trainUrl,
			keyTrain: self.config.ctaApiKeyTrain,
			idTrain: self.config.trainStationID,
			modInstance: self.config.moduleInstance
		};
		self.sendSocketNotification("CTA-REQUEST", request);  // Socket notification processed in node_helper.js;
	},

	getDom: function() {
		// console.log("Updating CTA dom"); // testing, make sure updateDom is running
		var self = this;
		// Create document container
		wrapper = document.createElement("div");
		// create table
		var table = document.createElement("table");

		if (this.dataNotification) {
			
			// Stop name header... create loop with bus row following later...
			// Check if bus is not null, if has data, run update:
			if (this.dataNotification.bus !== null) {
				var stopRow = document.createElement("tr");
				var stopRowElement = document.createElement("td");
				stopRowElement.align ="middle";
				stopRowElement.colSpan = "3";
				stopRowElement.className = "small";
				stopRowElement.innerHTML = this.config.busStopName;
				stopRow.appendChild(stopRowElement);
				table.appendChild(stopRow);

				// Do the bus title row
				var busRow = document.createElement("tr");
				busRow.className = "xsmall";
				busRow.align = "left";
				var dirElement = document.createElement("td");
				dirElement.innerHTML = "Direction";// dataNotification["bustime-response"].prd[0].rtdir;
				busRow.appendChild(dirElement);
				var rtElement = document.createElement("td");
				rtElement.align = "left";
				rtElement.innerHTML = "Route #"; // dataNotification["bustime-response"].prd[0].rt;
				busRow.appendChild(rtElement);
				var arrivalElement = document.createElement("td");
				arrivalElement.align = "right";
				arrivalElement.innerHTML = "Arrival" // dataNotification["bustime-response"].prd[0].rt;
				busRow.appendChild(arrivalElement);
				// Append busRow into table!
				table.appendChild(busRow);
			
				// Do the bus content row with a loop
				for (i = 0, len = this.dataNotification.bus["bustime-response"].prd.length; i < len; i++) {
					var arriveRow = document.createElement("tr");
					arriveRow.className = "xsmall";
					arriveRow.align = "left";
					var arriveElement = document.createElement("td");
					arriveElement.innerHTML = this.dataNotification.bus["bustime-response"].prd[i].rtdir;
					arriveRow.appendChild(arriveElement);
					var rtArriveElement = document.createElement("td");
					rtArriveElement.align = "left";
					rtArriveElement.innerHTML = this.dataNotification.bus["bustime-response"].prd[i].rt + " " + "<i class='fa fa-bus' aria-hidden='true'></i>";
					arriveRow.appendChild(rtArriveElement);
					var arrivalArriveElement = document.createElement("td");
					arrivalArriveElement.align = "right";
					arrivalArriveElement.innerHTML = this.dataNotification.bus["bustime-response"].prd[i].prdctdn + " min";
					arriveRow.appendChild(arrivalArriveElement);
					// Append busArrivalRow into table!
					table.appendChild(arriveRow);
				}
			};
			if (this.dataNotification.train !== null) {
				var stopRow = document.createElement("tr");
				var stopRowElement = document.createElement("td");
				stopRowElement.align ="middle";
				stopRowElement.colSpan = "3";
				stopRowElement.className = "small";
				stopRowElement.innerHTML = this.config.trainStopName;
				stopRow.appendChild(stopRowElement);
				table.appendChild(stopRow);

				// Do the bus title row
				var busRow = document.createElement("tr");
				busRow.className = "xsmall";
				busRow.align = "left";
				var dirElement = document.createElement("td");
				dirElement.innerHTML = "Direction"; // dataNotification["ctatt"].prd[0].rtdir;
				busRow.appendChild(dirElement);
				var rtElement = document.createElement("td");
				rtElement.align = "left";
				rtElement.innerHTML = "Route #"; // dataNotification["ctatt"].prd[0].rt;
				busRow.appendChild(rtElement);
				var arrivalElement = document.createElement("td");
				arrivalElement.align = "right";
				arrivalElement.innerHTML = "Arrival" // dataNotification["ctatt"].prd[0].rt;
				busRow.appendChild(arrivalElement);
				// Append busRow into table!
				table.appendChild(busRow);
			
				// Do the train content row with a loop
				for (i = 0, len = this.dataNotification.train["ctatt"].eta.length; i < len; i++) {
					var arriveRow = document.createElement("tr");
					arriveRow.className = "xsmall";
					arriveRow.align = "left";
					var arriveElement = document.createElement("td");
					arriveElement.innerHTML = this.dataNotification.train["ctatt"].eta[i].destNm;
					arriveRow.appendChild(arriveElement);
					var rtArriveElement = document.createElement("td");
					rtArriveElement.align = "left";
					rtArriveElement.innerHTML = "<i class='fa fa-subway' style='color:blue'></i>"; // Add options for other train colors
					arriveRow.appendChild(rtArriveElement);
					var arrivalArriveElement = document.createElement("td");
					arrivalArriveElement.align = "right";
					var arrivalT = moment(this.dataNotification.train["ctatt"].eta[i].arrT);
					var now = moment();
					var duration = moment.duration(arrivalT.diff(now));
					var arrivalTime = Math.round(duration.asMinutes());
					arrivalArriveElement.innerHTML = arrivalTime + " min";
					arriveRow.appendChild(arrivalArriveElement);
					// Append busArrivalRow into table!
					table.appendChild(arriveRow);
				}
			}
		};
		wrapper.appendChild(table);
		return wrapper;
	},

    	/* scheduleUpdate()
     	* Schedule next update.
     	* this.config.updateInterval is used.
     	* see:  https://github.com/nwootton/MMM-UKLiveBusStopInfo */
    	scheduleUpdate: function() {
        	var nextLoad = this.config.updateTime;
        	var self = this;
        	this.updateTimer = setInterval(function() {
			self.apiRequest(self);
		}, nextLoad);
    },

	socketNotificationReceived: function (notification, payload) {
		if (notification === this.instanceData) {
			// send payload (aka bus/train data to new var = dataNotification)
			console.log("Payload received");			
			this.dataNotification = payload;
			this.updateDom();
			// this.scheduleUpdate(this.config.updateTime); NO LONGER NEED SINCE USING setInterval()
		}
	},
});


node_helper.js
var request = require("request");
var NodeHelper = require("node_helper");

module.exports = NodeHelper.create({
	// Process socket notification from MMM-CTA.js

	start: function() {
		this.config = null;
		this.instance = null;
	},

	getData: function(payload) {
		var self = this;
		var busReady = false;
		var trainReady = false;
		this.instance = "MMM-CTA-DATA" + payload.modInstance;
		// Set myUrlBus based on what is set, first, if "top" is set, then if not set
		if (payload.urlBus !== null && payload.key !== null && payload.stpid !== null && payload.maxRes !== null) {
			var myUrlBus = payload.urlBus + "?key=" + payload.key + "&stpid=" + payload.stpid + "&top=" + payload.maxRes +  "&format=json";
			busReady = true;
		} else if (payload.urlBus !== null && payload.key !== null && payload.stpid !== null) {
			var myUrlBus = payload.urlBus + "?key=" + payload.key + "&stpid=" + payload.stpid + "&format=json";
			busReady = true;
		} else { console.log("MMM-CTA: Bus config incomplete!!!"); };
		
		if (payload.urlTrain !== null && payload.keyTrain !== null && payload.maxResTrain !== null && payload.idTrain !== null) {
			var myUrlTrain = payload.urlTrain + "?key=" + payload.keyTrain + "&max="+ payload.maxResTrain + "&mapid=" + payload.idTrain + "&outputType=json";
			trainReady = true;
		} else if (payload.urlTrain !== null && payload.keyTrain !== null && payload.idTrain !== null) { // Max result will be 4 by default
			var myUrlTrain = payload.urlTrain + "?key=" + payload.keyTrain + "&max=5" + "&mapid=" + payload.idTrain + "&outputType=json";
			trainReady = true;
		} else { console.log("MMM-CTA: Train config incomplete!!!"); };

		// Functions for bus & train, bus, or train only
		if (busReady && trainReady) {
			self.getBusTrain(myUrlBus, myUrlTrain);
		} else if (busReady) {
			self.getBus(myUrlBus);
		} else if (trainReady) {
			self.getTrain(myUrlTrain);
		} else { 
			console.log("No complete bus or train configs found!!!"); 
		};
	},	
	
	// console.log(myUrl); // for debugging
	getBusTrain: function(myUrlBus, myUrlTrain) {
		var self = this;
		console.log(self.instance);
		var bodyJs = {
			bus: null,
			train: null
		};
		request({url: myUrlBus}, function (error, response, body) {
			// console.log("CTA request fired."); // for debugging
			// Following line for building, delete when able to get DOM to show
			// self.sendSocketNotification("MMM-CTA-DATA", "TESTING");
			// Delete above when solved
			if (!error && response.statusCode == 200) {
				bodyJs.bus = JSON.parse(body);
				console.log(bodyJs.bus); // For testing purposes;
				request({url: myUrlTrain}, function (error, response, body) {
					console.log("CTA request fired train."); // for debugging
					// Following line for building, delete when able to get DOM to show
					// self.sendSocketNotification("MMM-CTA-DATA", "TESTING");
					// Delete above when solved
					if (!error && response.statusCode == 200) {
						bodyJs.train = JSON.parse(body);
						console.log(bodyJs.train); // For testing purposes;
						self.sendSocketNotification(self.instance, bodyJs)
					}	
				});
			} else {
				// If no data from bus or error from bus still run train!
				request({url: myUrlTrain}, function (error, response, body) {
					console.log("CTA request fired train."); // for debugging
					// If no error, store in bodyJs
					if (!error && response.statusCode == 200) {
					bodyJs.train = JSON.parse(body);
					console.log(bodyJs.train); // For testing purposes;
					self.sendSocketNotification(self.instance, bodyJs)
					};
				});
			}
		});
	},

	getBus: function(myUrlBus) {
		var self = this;
		var bodyJs = {
			bus: null,
			train: null
		};
		request({url: myUrlBus}, function (error, response, body) {
			// console.log("CTA request fired."); // for debugging
			// Following line for building, delete when able to get DOM to show
			// self.sendSocketNotification("MMM-CTA-DATA", "TESTING");
			// Delete above when solved
			if (!error && response.statusCode == 200) {
				bodyJs.bus = JSON.parse(body);
				self.sendSocketNotification(self.instance, bodyJs);
				console.log(bodyJs);
			};
		});
	},

	getTrain: function(myUrlTrain) {
		var self = this;
		var bodyJs = {
			bus: null,
			train: null
		};
		request({url: myUrlTrain}, function (error, response, body) {
			console.log("CTA request fired train."); // for debugging
			// If no error, store in bodyJs
			if (!error && response.statusCode == 200) {
				bodyJs.train = JSON.parse(body);
				console.log(bodyJs.train); // For testing purposes;
				self.sendSocketNotification(self.instance, bodyJs)
			};
		});
	},

	socketNotificationReceived: function(notification, payload) { // Payload is 'request' from MMM-CTA
		var self = this;
		if (notification === "CTA-REQUEST" ) {
			// console.log("Running CTA-REQUEST"); // for debugging			
			self.config = payload;
			self.getData(payload);
			}
	}

});


config.js
/* Magic Mirror Config Sample
 *
 * By Michael Teeuw http://michaelteeuw.nl
 * MIT Licensed.
 *
 * For more information how you can configurate this file
 * See https://github.com/MichMich/MagicMirror#configuration
 *
 */

var config = {
	address: "localhost", // Address to listen on, can be:
	                      // - "localhost", "127.0.0.1", "::1" to listen on loopback interface
	                      // - another specific IPv4/6 to listen on a specific interface
	                      // - "", "0.0.0.0", "::" to listen on any interface
	                      // Default, when address config is left out, is "localhost"
	port: 8080,
	ipWhitelist: ["127.0.0.1", "::ffff:127.0.0.1", "::1"], // Set [] to allow all IP addresses
	                                                       // or add a specific IPv4 of 192.168.1.5 :
	                                                       // ["127.0.0.1", "::ffff:127.0.0.1", "::1", "::ffff:192.168.1.5"],
	                                                       // or IPv4 range of 192.168.3.0 --> 192.168.3.15 use CIDR format :
	                                                       // ["127.0.0.1", "::ffff:127.0.0.1", "::1", "::ffff:192.168.3.0/28"],

	language: "en",
	timeFormat: 12,
	units: "imperial",

	modules: [
		{
			module: "alert",
		},
		{
			module: "updatenotification",
			position: "top_bar"
		},
		{
			module: "clock",
			position: "top_left"
		},
		{
			module: "calendar",
			header: "My Calendar",
			position: "top_left",
			config: {
				calendars: [
					{
						symbol: "calendar",
						maximumEntries: 5,
						url: "https://calendar.google.com/calendar/ical/nate.damaschke%40gmail.com/private-543bac710acf7b958c82b1d0a715cbbc/basic.ics",
					},
					{
						symbol: "calendar",
						maximumEntries: 3,
						url: "https://calendar.google.com/calendar/ical/en.usa%23holiday%40group.v.calendar.google.com/public/basic.ics",
					},

				]
			}
		},
		{
			module: "compliments",
			position: "bottom_bar",
			config: {
				updateInterval: 120000
			}
		},
/*		{
			module: "currentweather",
			position: "top_right",
			config: {
				units = "imperial",
				initialLoadDelay: 1000,
				location: "Chicago",
				locationID: "4887398",  //ID from http://www.openweathermap.org/help/city_list.txt
				appid: "182b84d55fe01275bde01f735ccd8b94"
			}
		},
		{
			module: "weatherforecast",
			position: "top_right",
			header: "Weather Forecast",
			config: {
				location: "Chicago",
				locationID: "4887398",  //ID from http://www.openweathermap.org/help/city_list.txt
				appid: "182b84d55fe01275bde01f735ccd8b94"
			}
		}, */
		
		{
			module: 'MMM-MyWeather',
			position: 'bottom_right',
			config: {
				apikey: '16f618a660e03f02', // private; don't share!
				pws: 'pws:KILCHICA206', //culemborg
				hourly: '0',
				fctext: '1',
				fcdaycount: "4",
				fcdaystart: "0",
				hourlyinterval: "3",
				hourlycount: "2",
				alerttime: 10000,
				alerttruncatestring: "english:",
				roundTmpDecs: 1,
				UseCardinals: 0,
				layout: "horizontal",
				sysstat: 0,
				coloricon: true
			}
		},
		{
			module: "newsfeed",
			position: "bottom_bar",
			config: {
				feeds: [
					{
						title: "New York Times",
						url: "http://www.nytimes.com/services/xml/rss/nyt/HomePage.xml"
					}
				],
				showSourceTitle: true,
				showPublishDate: true,
				updateInterval: 20000
			}
		},
		/* {
			module: "MMM-fitbit",
			position: "top_center",
			config: {
			credentials: {
				client_id: "22CPHW",
				client_secret: "c7e5b9e437c3cc1b54572d252c306c33",
				},
		resources: [
			'steps',
			'floors',
			'caloriesOut',
			'distance',
			'activeMinutes'
			],
		update_interval: 60
		}
		},
		{
			module: "MMM-fitbit2",
			position: "top_center",
			config: {
			credentials: {
				client_id: "22CPHT",
				client_secret: "569eb3a2b7e81c765e751226b7d3195b",
				},
		resources: [
			'steps',
			'floors',
			'caloriesOut',
			'distance',
			'activeMinutes'
			],
		update_interval: 60
		}
		}, */

		{
			module: "MMM-CTA",
			position: "bottom_left",
			config: {
					stopName: "Ashland and Blackhawk",
					stationId: null,
					stopId: 14619,
					maxResult: 5,
					routeCode: 66,
					ctaApiKey: 'fnuXHgKBkVcQmbAYXT9bWvCWF', //My api key: 'fnuXHgKBkVcQmbAYXT9bWvCWF',
					busUrl: 'http://www.ctabustracker.com/bustime/api/v2/getpredictions',
					updateTime: 60000
			}
		},
		{
		module: "MMM-MyScoreboard",
  		position: "top_right",
  		classes: "default everyone",
  		header: "My Scoreboard",
  		config: {
    			showLeagueSeparators: true,
    			colored: true,
    			viewStyle: "smallLogos",
			showRankings: true,
    			sports: [
      				{
        			league: "NCAAM",
        			groups: ["Big Ten"]
      				}]
			}
		}
	]
};

/*************** DO NOT EDIT THE LINE BELOW ***************/
if (typeof module !== "undefined") {module.exports = config;}
